File: Participación 4/Circulo2D.py
from multimethod import multimethod
import math
import numbers

class Circulo2D:
    def __init__(self, x=0, y=0, radio=1):
        self.x = x
        self.y = y
        self.radio = radio

    def get_area(self):
        return math.pi * self.radio * self.radio

    def get_perimetro(self):
        return 2 * math.pi * self.radio

    @multimethod
    def contiene(self, x: numbers.Real, y: numbers.Real):
        """Contiene un punto"""
        distancia = math.sqrt((self.x - x) ** 2 + (self.y - y) ** 2)
        return distancia <= self.radio

    @multimethod
    def contiene(self, c: "Circulo2D"):
        """Contiene otro círculo"""
        distancia = math.sqrt((self.x - c.x) ** 2 + (self.y - c.y) ** 2)
        return distancia + c.radio <= self.radio

    @multimethod
    def sobrepone(self, c: "Circulo2D"):
        """Sobrepone"""
        distancia = math.sqrt((self.x - c.x) ** 2 + (self.y - c.y) ** 2)
        return distancia <= (self.radio + c.radio) and distancia >= abs(self.radio - c.radio)
# -------------------------------
if __name__ == "__main__":
    c1 = Circulo2D(2, 0, 1)

    print("Área:", c1.get_area())
    print("Perímetro:", c1.get_perimetro())

    print("¿Contiene el punto (2.5,0)?", c1.contiene(2.5, 0))
    print("¿Contiene el círculo (2,0,0.5)?", c1.contiene(Circulo2D(2, 0, 0.5)))

    print("¿Se sobrepone con (0,0,2)?", c1.sobrepone(Circulo2D(0, 0, 2)))

File: Participación 4/P4_Java/P4_java/src/P4/Circulo2D.java
package P4;

public class Circulo2D {
    private double x;
    private double y;
    private double radio;

    public Circulo2D() {
        this.x=0;
        this.y=0;
        this.radio=1;
    }

    public Circulo2D(double x, double y, double radio) {
        this.x = x;
        this.y = y;
        this.radio = radio;
    }

    public double getX() { 
    	return x; 
    }
    public double getY() { 
    	return y; 
    }
    public double getRadio() { 
    	return radio; 
    }

    
    public double getArea() {
        return Math.PI * radio * radio;
    }

    public double getPerimetro() {
        return 2 * Math.PI * radio;
    }

    //contiene para un punto
    public boolean contiene(double x, double y) {
        double distancia = Math.sqrt(Math.pow(this.x - x, 2) + Math.pow(this.y - y, 2));
        return distancia <= radio;
    }

    //contiene para otro círculo
    public boolean contiene(Circulo2D c) {
        double distancia = Math.sqrt(Math.pow(this.x - c.x, 2) + Math.pow(this.y - c.y, 2));
        return distancia + c.radio <= this.radio;
    }

    //sobrepone
    public boolean sobrepone(Circulo2D c) {
        double distancia = Math.sqrt(Math.pow(this.x - c.x, 2) + Math.pow(this.y - c.y, 2));
        return distancia <= (this.radio + c.radio) &&
               distancia >= Math.abs(this.radio - c.radio);
    }
}

